src/CSnakes.Runtime/CPython/EventLoop.cs

```diff
@@ -285,32 +311,34 @@ private void RunForever()
                                 // If the task could not be scheduled, set the exception.
 
                                 request.CompletionSource.SetException(ex);
-                            }
-                            finally
-                            {
+
+                                // Clean up any resources allocated for the task.
+
+                                cancellationRegistration.Dispose();
                                 disposable?.Dispose();
                             }
                             break;
                         }
                         case (CancelRequest request, _):
                         {
-                            request.Task.Cancel(request.CancellationToken);
+                            var scheduled = scheduledTasks.Find(t => ReferenceEquals(t.Task, request.Task));
+                            scheduled?.Cancel(request.CancellationToken);
                             break;
                         }
                         case (StopRequest, RunState.Running):
                         {
                             state = RunState.Stopping;
-                            foreach (var task in coroTasks)
-                                task.Cancel();
+                            foreach (var task in scheduledTasks)
+                                task.Cancel(default);
```

Comment:
nit: Make it more explicit:

```suggestion
                                task.Cancel(CancellationToken.None);
```